doc/asynchronous_fifo/asynchronous_fifo.rst

```diff
@@ -0,0 +1,463 @@
+Asynchronous FIFO
+=================
+
+An Asynchronous FIFO is a non-blocking data structure in which elements
+gets pushed in on one side and pulled out on the other side. The key to
+this component is the non-blocking nature, and the underlying assumption
+that software on both sides rate-matches their requests. The asynchronous
+FIFO has a PID control inside it that can be used to control either the
+producer or the consumer.
+
+Two typical use cases are shown in :ref:`asycnhronous_FIFO_use_cases`.
+In te first use case we have an Asynchronous Sample Rate Converter (ASRC) in
+front of the FIFO. The task of this ASRC is to dynamically introduce or
+remove samples in order to match the rate of producer and consumer. In the
+second use case we have a PLL (either hardware or software) that is used to
+match the rate of the producer and consumer.
+
+.. _asycnhronous_FIFO_use_cases:
+
+.. figure:: images/use-cases.*
+            :width: 75%
+
+            Use cases for the asynchronous FIFO
+
+In order to use the asynchronous FIFO one needs at least two threads: a
+producer thread (on the left), and a consumer thread (on the right). These
+threads are free running relative to each other, and the FIFO transports
+data from the producer to the consumer. The FIFO has a fixed length, set on
+creation, and the control algorithm inside the FIFO tries and keep the FIFO
+half-full at all times. When the producer is slower than the consumer the
+FIFO will drain a bit until the rates match again, and when the producer is
+faster than the consumer the FIFO will grow until the rates match again. In
+order to ensure that the FIFO stays half full, the control algorithm will
+always slightly overshoot on a a step-change in rates.
+
+Note that the FIFO is unaware whether it is the producer that is too fast,
+or the consumer that is too slow. It does not attribute blame for a
+rate-mismatch. The FIFO just observes the mismatch. 
+
+
+Using the Asynchronous FIFO
+---------------------------
+
+An Asynchronous FIFO is allocated as an array of double-word integers::
+
+  int64_t array[ASYNCHRONOUS_FIFO_INT64_ELEMENTS(ENTRIES, SAMPLE_SIZE)];
+
+The ``ASYNCHRONOUS_FIFO_INT64_ELEMENTS()`` macro calculates the number of
+double words required for the FIFO given the number of entries in the FIFO,
+and the number of words that each sample occupies. For example, when
+transferring stereo Audio through a fifo with 40 elements one would use
+``ASYNCHRONOUS_FIFO_INT64_ELEMENTS(40, 2)``. Note that the two elements are
+not interchangeable.
+
+The number of elements in the FIFO is a trade-off that the system designer
+makes. As the FIFO will always aim to be half-full, a large number of
+elements will introduce a high latency in the system and occupy a large
+amount of memory. A short FIFO wil contribute little latency but may easily
+overflow and underflow.
```

Comment:
I wonder if a note about FIFO sizing makes sense? I found that 2x the biggest input block size wasn't enough. 3x covered it fine for all SR changes across the matrix.